### DISCUSSION TITLE
Support more complex repository skip patterns

### DISCUSSION THREAD

**Discussion by luxaritas** [lowlighter/metrics#1217] [Ideas]
My primary use case here would be to skip all repositories under some organization like `MyOrg/*`, however I could imagine it being useful in some cases to match other patterns (eg if there is a set of repositories with a similar naming scheme that should be skipped)

**Comment by luxaritas** [comment]
Worth noting: With something like this it would be good to also have a way to revert a pattern, eg: `MyOrg/*, !MyOrg/SomeRepo`

**Reply by spenserblack** [reply]
Sounds cool! A couple of thoughts:
1. This looks a lot like an ignorefile syntax (`.gitignore`, `.npmignore`, etc.), so there should be a library that can be used for this project
2. If it uses an ignorefile's syntax, maybe it should be possible to make it a separate file from the workflow?
   ```yaml
   # metrics.yml
   # instead of this
   repositories_skipped: MyOrg/*, !MyOrg/SomeRepo, Me/SkippedRepo
   ```
   ```ignore
   # .metricsignore
   # This file can be read by the workflow as long as it's in the same repo as the workflow
   MyOrg/*
   !MyOrg/SomeRepo
   ```

**Reply by lowlighter** [reply]
Yeah this looks good indeed 👍 
Plus currently skip are handled by each plugin so it'll help in maintainability and predictability to use a single shared filter that respect a given syntax (just need to be backwards compatible with current syntax though)

_____

As for a `.metricsignore`, I'd like to avoid "magic files" because usually it's a bit trouble to handle depending on the context, but we could get a similar behaviour if the above syntax supported remote file urls,  something like `https://github.com/repo/user/blob/master/.metricsignore` (seems more flexible too)

**Reply by lowlighter** [reply]
Not promising anything, but it may be feasible to also include basic criteria filtering too (like ignoring forks, less than x stars, not updated since 2y, etc.)

If this is something that you feel like that are missing too we could eventually check to add it in the meantime

**Reply by spenserblack** [reply]
If `repositories_skipped` was made newline-separated, then an ignorefile syntax could easily be embedded right in the config.
```yaml
repositories_skipped: |
  username/exact-match
  org/matches-multiple-*
  !org/matches-multiple-except-this
```
Or it doesn't even need to be newline-separated, you could probably just take the text as-is and feed it to an ignorefile parser.

The above example is a breaking change, but that can be avoided if `repositories_skipped` is kept the same (maybe trigger a deprecation warning if the input isn't empty), and an input with a different name is used instead.

**Comment by lowlighter** [comment]
This is currently in testing on master/main branches 👍 

Quick example:
```yml
repositories_skipped: |
  @use.patterns
  
  # Skip a specific repository (both patterns are equivalent)
  user/repo 
  -user/repo

  # Skip repositories matching a given pattern
  user/repo-*
  {user1, user2, user3}/*
  
  # Include back a previously skipped repository
  org/repo
  +org/include-this-repo
```

The pattern matching is a file-glob flavor, and operators `-`/`+` can be used to respectively exclude/include repos (`-` is implicit and can be ommitted)

See [`🔕 Skipping repositories in plugins`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#-skipping-repositories-in-plugins) for more informations

If you have any feedback, please feel free to share them 😄 !
